backend/src/api/song_routes.py

```diff
@@ -0,0 +1,179 @@
+    if num_likes > 0:
+        data["num_likes"] = num_likes
+
+    return data
+
+
+@song_routes.post("")
+@login_required
+def upload_song() -> ApiErrorResponse | tuple[IdAndTimestamps, int]:
+    """
+    Create a new record of a song on the Songs table.  Redirect user to that song's page
+    """
+    form = SongForm()
+    form["csrf_token"].data = request.cookies["csrf_token"]
+    if form.validate_on_submit():
+        new_song = Song(
+            name=form.data["name"],
+            artist_id=current_user.id,
+            genre=form.data["genre"],
+            thumb_url=form.data["thumb_url"],
+            song_ref=form.data["song_ref"],
+            created_at=datetime.now(timezone.utc),
+            updated_at=datetime.now(timezone.utc),
+        )
+        db.session.add(new_song)
+        db.session.commit()
+
+        song_data: IdAndTimestamps = {
+            "id": new_song.id,
+            "created_at": str(new_song.created_at),
+            "updated_at": str(new_song.updated_at),
+        }
+        return song_data, 201
+
+    return form.errors, 400
+
+
+@song_routes.put("/<int:song_id>")
+@login_required
+def update_song(song_id: str) -> ApiErrorResponse | IdAndTimestamps:
+    """
+    Change data about a song that exists as long as song exists and current_user is authorized to do so
+    """
+    id = int(song_id)
+    song_to_update = db.session.query(Song).filter(Song.id == id).one_or_none()
+
+    if not song_to_update:
+        return song_not_found_error
+
+    if song_to_update.artist_id != current_user.id:
+        return not_authorized_error
+
+    form = SongForm()
+
+    form["csrf_token"].data = request.cookies["csrf_token"]
+    if form.validate_on_submit():
+        song_to_update.name = form.data["name"]
+        song_to_update.genre = form.data["genre"]
+        song_to_update.thumb_url = form.data["thumb_url"]
+        song_to_update.song_ref = form.data["song_ref"]
+        song_to_update.updated_at = datetime.now(timezone.utc)
+
+        db.session.commit()
+
+        song_data: IdAndTimestamps = {
+            "id": song_to_update.id,
+            "created_at": str(song_to_update.created_at),
+            "updated_at": str(song_to_update.updated_at),
+        }
+
+        return song_data
+
+    return form.errors, 400
+
+
+# I want to be able to delete a song that I uploaded
+@song_routes.delete("/<int:song_id>")
+@login_required
+def delete_song(song_id: str) -> DeleteSong | ApiErrorResponse:
+    """
+    Delete a song as long as the song exists and the current_user is authorized to do so
+    """
+    id = int(song_id)
+    song_to_delete = db.session.query(Song).filter(Song.id == id).one_or_none()
+
+    if not song_to_delete:
+        return song_not_found_error
+
+    if song_to_delete.artist_id != current_user.id:
+        return not_authorized_error
+
+    db.session.delete(song_to_delete)
+    db.session.commit()
+
+    return {}  # can we avoid returning this empty dictionary?
```

Comment:
ive actually done the same, flask doesn't seem very happy with a None so maybe we could change this to a string and DeleteSong to a `Literal[""]` ?

Reply:
Since you're doing the same, I'm going to leave it for now.  Seems like something we could refactor later